### PR TITLE
feat: add vertexConstantAttributeBuilder

### PR DESCRIPTION
**Pull request by josted**
No sign of ANGLE_instanced_arrays yet. 

Still to-do for this version:
- [x] - [Update with master again] 
- [x] - [Hook up to candlestick] 
- [x] - [And ohlc] 
- [x] - [And boxplot] 
- [x] - [Pull base(gl, program, name) call to above validSize check] 

Partly addresses https://github.com/d3fc/d3fc/issues/1369

### CODE REVIEW

**Comment by jamesmporter** [comment]
Does this replace the `projectedAttributeBuilder`? Should it therefore delete that file?

File: packages/d3fc-webgl/src/buffers/vertexConstantAttributeBuilder.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+                target++;
+            }
+        }
+        for (
+            let index = elementComponentCount;
+            index < projectedData.length;
+            index += elementComponentCount
+        ) {
+            projectedData.copyWithin(
+                index,
```
Comment (chrisprice) [inline packages/d3fc-webgl/src/buffers/vertexConstantAttributeBuilder.js]: Should this be `offset + index`?

File: packages/d3fc-webgl/src/buffers/vertexConstantAttributeBuilder.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        }
+        let target = 0;
+        for (let vertex = 0; vertex < verticesPerElement; vertex++) {
+            if (data[vertex].length !== components) {
+                throw new Error(
+                    `Expected components array of size ${components}, recieved array with length ${data[vertex].length}.`
+                );
+            }
+            for (let component = 0; component < components; component++) {
+                projectedData[target] = data[vertex][component];
```
Comment (chrisprice) [inline packages/d3fc-webgl/src/buffers/vertexConstantAttributeBuilder.js]: Should this be `projectedData[offset + target]`?

File: packages/d3fc-webgl/src/buffers/vertexConstantAttributeBuilder.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+export default () => {
+    const base = baseAttributeBuilder();
+
+    let data = null;
+    let validSize = 0;
+
+    const project = (elementCount, verticesPerElement) => {
+        const components = base.size();
+        const offset = base.offset();
+        const elementComponentCount = verticesPerElement * components;
```
Comment (chrisprice) [inline packages/d3fc-webgl/src/buffers/vertexConstantAttributeBuilder.js]: `componentsPerElement`?

**Review by chrisprice — CHANGES_REQUESTED** [review]
I agree with @jamesmporter's comment, if we can delete `projectedAttributeBuilder` as part of this then we should

File: packages/d3fc-webgl/src/buffers/vertexConstantAttributeBuilder.js (hunk 3 above)
Comment (josted) [reply packages/d3fc-webgl/src/buffers/vertexConstantAttributeBuilder.js]: Sounds good.

File: packages/d3fc-webgl/src/buffers/vertexConstantAttributeBuilder.js (hunk 1 above)
Comment (josted) [reply packages/d3fc-webgl/src/buffers/vertexConstantAttributeBuilder.js]: Done.

File: packages/d3fc-webgl/src/buffers/vertexConstantAttributeBuilder.js (hunk 2 above)
Comment (josted) [reply packages/d3fc-webgl/src/buffers/vertexConstantAttributeBuilder.js]: I reckon so, done.

**Comment by josted** [comment]
> I agree with @jamesmporter's comment, if we can delete `projectedAttributeBuilder` as part of this then we should

I think I'll need to wait for box-plot to go in, but should be straightforward enough after that.

File: packages/d3fc-webgl/src/series/glArea.js [hunk 4]
```diff
@@ -32,9 +33,8 @@ export default () => {
             [0, 0, 1],
             [1, 0, 0],
             [1, 1, 0]
-        ])
-        .value((data, element, vertex, component) => data[vertex][component]);
-    const definedAttribute = projectedAttributeBuilder().value(
+        ]);
+    const definedAttribute = elementConstantAttributeBuilder().value(
```
Comment (josted) [inline packages/d3fc-webgl/src/series/glArea.js:36]: Should this definitely be elementConstantAttribute builder here? It appears to work.

File: packages/d3fc-webgl/src/series/glLine.js [hunk 5]
```diff
@@ -28,16 +29,15 @@ export default () => {
...
         .data([
             [-1, -1],
             [1, -1],
             [-1, 1],
             [1, 1]
-        ])
-        .value((data, element, vertex, component) => data[vertex][component]);
-    const definedAttribute = projectedAttributeBuilder().value(
+        ]);
+    const definedAttribute = elementConstantAttributeBuilder().value(
```
Comment (josted) [inline packages/d3fc-webgl/src/series/glLine.js:39]: Same question as above - both elementConstant and vertexConstant display the graph as intended but not sure of the 'invisible' implications of either one here.

File: packages/d3fc-webgl/src/series/glArea.js (hunk 4 above)
Comment (chrisprice) [reply packages/d3fc-webgl/src/series/glArea.js:36]: Yep as discussed.

**Comment by chrisprice** [comment]
Blocked on #1371.

**Comment by chrisprice** [comment]
FYI #1371 is now done, this PR should now be unblocked.

**Review by chrisprice — APPROVED** [review]

**Pull request merged**